### ISSUE TITLE
Mouse cursor should reflect which parts of the UI are clickable

### ISSUE THREAD

**Issue by eloquence** [freedomofpress/securedrop-client#415]
Quick iconographic reference for mouse cursors (exact icons will vary based on OS):

- Arrow pointer state: ![](https://raw.githubusercontent.com/daviddarnes/mac-cursors/master/svg/pointer.svg?sanitize=true) 
- I-beam state: ![](https://raw.githubusercontent.com/daviddarnes/mac-cursors/master/svg/ibeam.svg?sanitize=true) 
- Hand pointer state: ![](https://raw.githubusercontent.com/daviddarnes/mac-cursors/master/svg/pointinghand.svg?sanitize=true)

Currently, the mouse cursor is in arrow state throughout, with the exception of the reply area and the inputs on the login dialog, where it is in I-beam state. 

The mouse cursor should be set to hand pointer state  for all clickable areas of the UI that are not typing areas:

- all buttons, including the "Send Reply" button
- individual source list entries
- the clickable area of the AuthWidget
- the "Sign out" button, if expanded

# Acceptance Criteria

**Given** that I am running the SecureDrop client
**When** I move my mouse over an area that is clickable, but within which I cannot type
**Then** the mouse cursor should change into "hand pointer" state
**And** it should change to the appropriate new state if I leave this area of the UI

**Given** that I am running the SecureDrop client
**When** I move my mouse over an area that is clickable, and within which I can type
**Then** the mouse cursor should change into "I-beam" state
**And** it should change to the appropriate new state if I leave this area of the UI

# User Story

As a user, I want to be able to quickly identify parts of the UI I can interact with by moving my mouse over them, so that I can navigate the user interface in familiar ways.

**Comment by ninavizz** [comment]
## Zeplin spec for hover + hover w/in hover
- **Screen:** https://scene.zeplin.io/project/5c807ea562f734bd2756b243/screen/5cfeec7d628fd215c7258bb9
- Notes
  - Unless otherwise specified, whenever the item is hovered over it should get the pointer-hand. All but 2 items on this spec hover over actionable things, so this is no different than Erik's point in the OG comment.
  - Pink dotted-lines = approximated hit/target areas
  - Pls see authenticated Zeplin screen to grab CSS values on dropshadows & other color specifics
- **0:** Refresh widget does not change colors nor get a dropshadow, on hover
- **1:** Username that is truncated is revealed, upon hover; on click, it shows the "Widget Open" view (that reveals Sign Out option)
- **2:** Star w/in Conversation Pane header; gets a white dropshadow, gets a 2px stroke, white fill
- **3:** More icon; gets a white dropshadow, does not change colors
- **4:** Source w/in list: Text goes from all being `Charcoal`, to black; paper-clip, same
- **5:** Delete w/in Selected Source: Text/Icon both go from `Midnite Blue` to `Grimace Blue`, both get white dropshadows
- **6:** Journalist Badge w/in Conversation Pane: Cursor does NOT go to pointer-hand, as there is nothing to click on. When badge is hovered-over tho, the full Fn/Ln in Sentence Case both show against same colored background as badge. 
  - If Vyachaslav Grebenshikov is the authenticated user, the name on the flyout would not be his own but would instead say `You!`
- **7:** All of the Files Pane stuff is in its own spec, though this version is unlikely to make it into Beta.
- **8:** Star w/in Source w/in Source List: Fill goes to White, gets a 2px inner-stroke, and a cyan dropshadow.
- **9:** Airplane goes pink when hovered over
- **10:** Anywhere the user mouses-into the Reply pane, mouse goes to I-Beam

**Comment by eloquence** [comment]
@ninavizz recommendations from #574:

_State names based on [Mozilla's handy-dandy guide](https://developer.mozilla.org/en-US/docs/Web/CSS/cursor)_
**On Active item hover**
  - Whenever a user hovers their cursor over an active item, it should transition from `default` to `pointer` (an arrow to the mickey-mouse gloved pointy finger hand)
**On text-area hover _or_ click-into**
  - Whenever a user clicks into the _Reply_ pane (or any other dealie-o that lets them edit things) their cursor should transition from `default` to `text` (an arrow to the i-bar)
**STRETCH GOAL: Progress/Wait cursor indicators**
  - Whenever a user invokes any of the below _non-modal actions_, the cursor should transition from `default` to `progress`. Upon the activated task's completion, the cursor should return to `default`.
     - Opening a file
     - Checking the server for updates (manually invoking sync)
  - Whenever a user invokes any of the below _non-modal actions_, the cursor should transition from `default` to `wait`. When the activated task has been completed and the rest of the screen may be interacted with, the cursor should return to `default`. 
     - None, as of now. Export/Print should not do this, as that would be frightfully obnoxious.

**Comment by ntoll** [comment]
From a dev point of view the `my_widget.setCursor(QCursor(QtCore.Qt.PointingHandCursor))` is really all that's required. Now Qubes is happily installing on my sacrificial laptop, I'm on this. :+1:

**Comment by ntoll** [comment]
Hi folks. See the upcoming PR.

TL;DR it's mostly done given the caveats explained below (generally down to Qt limitations). As always feedback, constructive critique and ideas are most welcome. :+1: 

Implementation notes, numbered to follow @ninavizz's numbering system (for the UI in https://scene.zeplin.io/project/5c807ea562f734bd2756b243/screen/5cfeec7d628fd215c7258bb9), follows:

0. Done.
1. It's not possible to reveal the full length of the username widget on hover. However, I've added some code to check the length of the displayed name and if greater than the width of the widget, the username is set as the tooltip for the widget, so, on hover, you get to see the full untruncated name.  I hope this is an acceptable work-around. Also, while it's possible to set the pointing hand cursor to the username widget, when the menu (containing the single "SIGN OUT" item) is activated, then the cursor image for that widget is displayed while moving the mouse around the whole application window -- I presume because the menu is the widget with focus. Ergo, I've not set the cursor for the menu (it reverts to an arrow). Again, it's easy to change, but just highlighting my "executive" decision given the limitations of the Qt framework in this respect.
2. There is no such star in the current state of the GUI that I can see. Ignoring.
3. This works in the same way as (1), so the same cursor related limitations apply here (on hover you get the hand, when activated then whatever the cursor is for that menu is used all over the app window because the menu is active). As always, tell me how you want to proceed. :-)
4. I wasn't able to work out how to change the text colour on hover (believe me, I tried!). I hope you don't mind, but I resolved the situation by changing the background to a very light grey on hover. The issue is simply that the parent widget representing the whole item in the list is the thing we're checking for hovering, but the update to `color: #666;` (or whatever colour we use) wasn't happening because the child widgets (labels mostly) had CSS attached to them which it appears overrides anything from the parent widget. :-/
5. There is no such star in the current state of the GUI that I can see. Ignoring.
6. There is no such item in the current state of the GUI that I can see. Ignoring.
7. Care to link to the spec..? ;-)
8. There's no such thing as a drop shadow in Qt's version of CSS. I also had a play with the `QGraphicsDropShadowEffect` class but it didn't appear to work. Ergo, I've "hacked" it by setting a rounded border of 4px set to unselected-star-blu on `:hover`. I originally used cyan, but it didn't look very "drop shadow" and felt quite strong. Also, I ignored the fill instruction since that's what is used to flag if the item is starred or not (i.e. I imagine we don't want to un-fill a filled star on hover, since we're appearing to change the state of the flag when, in fact, no such thing is taking place).
9. Oh boy, I spent ages trying to work out how to make this work so there's some sort (any sort!) of hover effect on the send button. The CSS way to update the image appears not to work (no matter how hard I've tried various tricks). The other option would be to completely subclass the QPushButton class and reimplement a HoverableButton class, which has two problems: 1) it's very overkill for a simple effect, 2) Surely such behaviour should be found within the CSS aspects of the code rather than hard coded in. In the end (and I can't apologise enough for this), I got fed up and went for something which fitted your intention rather than specific design description, and made it behave like 8 (on hover).
10. Done.

@eloquence I've also made a start on implementing the wait-cursor when the user clicks the "sync" button (top left). This pattern / approach could be used elsewhere (such as when opening a file).

**Comment by eloquence** [comment]
Unfortunately we only today realized that Qubes does not support changing the mouse cursor, see discussion in https://github.com/QubesOS/qubes-issues/issues/1551. While we'd love to see a resolution to that in future versions of Qubes (if it can be done in a responsive and secure manner), it means that we can't rely on mouse cursor changes for indicating whether a UI element is clickable.

We've agreed that we can make these changes regardless, optimistically, but given current state, this significantly increases the priority of #591 (to give the user a clear indication which UI elements are clickable).

**Comment by ntoll** [comment]
@eloquence ack.

Importantly, some of the work in this PR *does* give a graphical indication of which elements are clickable via hover-related changes (e.g. change of colour), rather than a change to the mouse pointer.

**Comment by ninavizz** [comment]
@ntoll I think this was where the icky shadowing stuff came up... sorry to confuse! :/
